File: 2015/day20/day20.hs
import Debug.Trace
import Data.Map (Map)
import qualified Data.Map as Map
import Data.List

update_divisors current next =
    let
        first = head $ dropWhile (\x->(next`mod`x/=0)) [2..(floor $ sqrt $ fromIntegral next)]++[next]
    in
        if first == next then Map.insert next (1+next) current
        else
            let
                exps = [i | i<-[0..ceiling $ logBase 2 (fromIntegral next)+1], next`mod`(first^i)==0]
                power_sum = sum [first^i | i<-exps]
                divided = next `div` (first^(last exps))
            in
                Map.insert next (((Map.!) current divided)*power_sum) current
        
small_sum n = sum [n`div`x | x<-[1..50], n`mod`x == 0]

input = 6000--input goes here
target_one = input `div` 10
target_two = input `div` 11

damn_son = foldl' update_divisors (Map.fromList [(1,1)]) [2..target_one]
part_one = head $ dropWhile (\x-> ((Map.!) damn_son x < target_one)) [1..target_one]
part_two = head $ dropWhile (\x-> (small_sum x < target_two)) [1..target_two]

File: 2015/day20/day20.py
import math
input_num = 6000#input goes here
target = input_num//10
target_two = input_num//11

div_sum = {1:1, 2:3}

for i in range(3, target):
   if i%10000==0:print(i)
   p = [x for x in range(2, math.ceil(math.sqrt(i))+1) if i%x==0]
   if p==[]:
       div_sum[i] = i+1
   else:
       exps = [j for j in range(math.ceil(math.log2(i))+1) if i%(p[0]**j) == 0]
       div_sum[i] = div_sum[i//((p[0])**(exps[-1]))]*(sum([(p[0]**x) for x in exps]))
   if div_sum[i]>target:
       print("part one: ", i)
       break       
    

for i in range(2, target_two):
    new_sum = sum([i//x for x in range(1, 51) if i%x==0])
    if new_sum > target_two:
        print("part two: ", i)
        break